Microsoft.Toolkit.HighPerformance/Buffers/StringPool.cs

```diff
@@ -0,0 +1,475 @@
+            {
+                lock (this.dummy)
+                {
+                    ref string?[]? entries = ref this.entries;
+
+                    entries ??= new string[entriesPerBucket];
+
+                    int entryIndex = hashcode & (this.entriesPerBucket - 1);
+
+                    ref string? entry = ref entries.DangerousGetReferenceAt(entryIndex);
+
+                    if (!(entry is null) &&
+                        entry.AsSpan().SequenceEqual(value.AsSpan()))
+                    {
+                        return entry;
+                    }
+
+                    return entry = value;
+                }
+            }
+
+            /// <summary>
+            /// Implements <see cref="StringPool.GetOrAdd(ReadOnlySpan{char})"/> for the current <see cref="Bucket"/> instance.
+            /// </summary>
+            /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> with the contents to use.</param>
+            /// <param name="hashcode">The precomputed hashcode for <paramref name="span"/>.</param>
+            /// <returns>A <see cref="string"/> instance with the contents of <paramref name="span"/>, cached if possible.</returns>
+            public string GetOrAdd(ReadOnlySpan<char> span, int hashcode)
+            {
+                lock (this.dummy)
+                {
+                    ref string?[]? entries = ref this.entries;
+
+                    entries ??= new string[entriesPerBucket];
+
+                    int entryIndex = hashcode & (this.entriesPerBucket - 1);
+
+                    ref string? entry = ref entries.DangerousGetReferenceAt(entryIndex);
+
+                    if (!(entry is null) &&
+                        entry.AsSpan().SequenceEqual(span))
+                    {
+                        return entry;
+                    }
+
+                    // ReadOnlySpan<char>.ToString() creates a string with the span contents.
+                    // This is equivalent to doing new string(span) on Span<T>-enabled runtimes,
+                    // or to using an unsafe block, a fixed statement and new string(char*, int, int).
+                    return entry = span.ToString();
+                }
+            }
+
+            /// <summary>
+            /// Implements <see cref="StringPool.TryGet"/> for the current <see cref="Bucket"/> instance.
+            /// </summary>
+            /// <param name="span">The input <see cref="ReadOnlySpan{T}"/> with the contents to use.</param>
+            /// <param name="hashcode">The precomputed hashcode for <paramref name="span"/>.</param>
+            /// <param name="value">The resulting cached <see cref="string"/> instance, if present</param>
+            /// <returns>Whether or not the target <see cref="string"/> instance was found.</returns>
+            public bool TryGet(ReadOnlySpan<char> span, int hashcode, [NotNullWhen(true)] out string? value)
+            {
+                lock (this.dummy)
```

Comment:
Are these locks actually necessary? What occurs if they aren't here? Given a string write is atomic

Reply:
The `Bucket` type itself is thread-safe (each bucket acts as a lock, so we distribute the concurrency contentions across the existing buckets), and that's to synchronize access to the pool in each bucket. Without some sort of lock, we could have all kinds of race conditions ranging from eg. a bucket being reset after another thread has checked it for `null` (so we'd get an NRE), to race conditions potentially leading to multiple allocations, etc.

I mean sure assigning a `string` to a field is indeed atomic, but we're also doing some additional operations when setting/getting a pooled instance that still need to belong to the same critical section and be thread-safe to avoid issues.

Are you worried about anything in particular here? The thread contention should be very low here since we're subdividing into multiple buckets, so I'm not really worried about that being a bottleneck here 🤔

Reply:
:+1: All good then